### DISCUSSION TITLE
Worker from Class

### DISCUSSION THREAD

**Discussion by Fancy2209** [airsdk/Adobe-Runtime-Support#3305] [Ideas]
Currently, this requires third party libraries that pretty much involve some Low Level Dark Magic with a library called AS3SWF, but it'd be nicer to just be able to make a new worker from a class directly.
is there any chance AIR could have this natively?

**Comment by ajwfrost** [comment]
The way workers work is to essentially set up a secondary (albeit slimmed down) runtime in a separate thread, and it initialises the 'player' object using a SWF file format.. so, at some point, there needs to be the 'dark magic' for this. But potentially we could provide this natively i.e. provide a new AS3 method to create a Worker from a Class and then we'd craft the rest of the SWF format around this.. although this would imply that the given Class has no dependencies on anything else?

Are you thinking of this for fairly simple bits of async functionality that can be wholly encapsulated into a single Class? There's a danger that you could pass in a class that has dependencies on other (non-builtin) classes and then the worker would just fail to run..

thanks

**Reply by Fancy2209** [reply]
I was thinking it more to just, not need to wrap the main fuction in a `if (Worker.current.isPrimordial)` check, not beeing able to have external dependencies would be annoying as Physics would be one of the uses I wanted to use workers for, and I don't want to code a physics engine from scratch, specially in a single Class

**Reply by ajwfrost** [reply]
Hmmm... so how about if we could specify what classes were used as the 'root' for different Workers?
- With the primordial/first one, we spin up the normal 'main' class per the SWF definition
- but we could then also automatically relaunch the SWF as a bunch of Workers each of which is launched with a different class definition - taken from this same SWF so that we know (from the compiler) that all dependencies should be resolved already

Maybe via the app descriptor:
```
  <Workers>
    <Worker name="testWorker" startAtLaunch="true">com.test.workers.MyFirstWorker</Worker>
    <Worker name="customWorker" startAtLaunch="false">com.test.workers.OtherWorker</Worker>
  </Workers>
```
then we would trigger some of them to start automatically, and some would need to be kicked off via `Worker.start()`, but all would be available via `WorkerDomain.current.getWorker("name")` ..

The primordial, and each of the other workers, would be launched using the specified class name, and you don't need to do any checks within the primordial's constructor (unless of course you gave the same class name for one of these Worker entries!)

Would that be a suitable method?

We could still support the idea of creating a worker from a class name, within the main SWF, so just `worker = WorkerDomain.current.createFromClass(com.test.workers.MyFirstWorker)` which would reuse the same SWF (to avoid dependencies issues) but would trigger the start-up to use that class definition rather than the normal 'main' one...

**Reply by Fancy2209** [reply]
These ideas sound great! Would definitely make using Workers easier to use.

**Reply by Fancy2209** [reply]
And that last idea is good for libraries using workers for multithreading

**Reply by bobaoapae** [reply]
@ajwfrost hey, this idea it's amazing, will help a lot to use multiple workers to to differents jobs. I have a use for this already. Can we expect some beta build with this soon?

**Reply by ajwfrost** [reply]
That's in the plan for AIR 52.. around the turn of the year I would hope!

**Comment by rdefalco** [comment]
Hi guys,
I don't know if we can help, it's a thing we investigated many years ago.

We happen to have a class "WorkerManager" in the package "com.myflashlabs.utils.worker" that we invoke like this to create a new thread:

`downloadManager = new WorkerManager(workers.DownloadWorker, OperazioniPianificateCatalogo.getInstance().getLoaderInfo().bytes, this);`

it's a class that we inherited when we had an all-in subscription to MyFlashLabs ANEs. I don't know if these classes were part of the subscription, but I see "The MIT License (MIT)" in the header comments of all these classes.

Also in the comments I read that these classes are based on the work of:
//	https://github.com/bortsen/worker-from-class
//	https://github.com/claus/as3swf

**Comment by RayGyoe** [comment]
The AS3 runtime is designed for a single thread, making the implementation of workers extremely costly, unlike any other language. Of course, other programming paradigms have started to embrace coroutines...

**Reply by Fancy2209** [reply]
Workers already exist, they run in a separate, trimmed down AVM2 in a separate thread

**Reply by ajwfrost** [reply]
Yes although both of these are true -> each Worker is a complete instance of the AS3 (avmplus) runtime, but without most of the 'extra' bits that make it into an AIR runtime.

I have often wondered why it's not possible to do more traditional multithreading .. I think it's mostly just related to the memory management and garbage collection. Could be an interesting experiment to play with that at some point! But anyway - this is similar to the JavaScript engines in browsers or in Node, which are also single-threaded but with better asynchronous mechanisms so that Promises etc can be implemented natively.
